src/units/abilities.cpp

```diff
@@ -1311,7 +1311,16 @@ namespace { // Helpers for attack_type::special_active()
 			return false;
 		}
 
-		unit_filter ufilt{vconfig(*filter_child)};
+		//'inform' u.matches what called from a weapon special if filter_child contain "has_attack" child.
+		config cfg = *filter_child;
+		if ((*filter_child).has_child("has_attack") ) {
+			config& filter_has_attack = cfg.mandatory_child("has_attack");
+			filter_has_attack["weapon_type"] = tag_name;
```

Comment:
Adding `filter_has_attack["weapon_type"] = "damage_type"` seems confusing, given the variable name I didn't expect the string to be "damage_type". I think `filter_has_attack["bug_8940"] = tag_name` would be more understandable, as it's immediately obvious that it's a workaround.

I don't like the mechanism, and think it's the wrong solution for the master branch. However, assuming that a better fix will be so large that it nothing in this area gets backported from master to 1.18, it makes sense to fix 1.18 this way.

Reply:
Name changed